bgzf/cache.go

```diff
@@ -126,20 +125,30 @@ func (b *block) Read(p []byte) (int, error) {
 	return n, err
 }
 
-// readToEOF will exhaust the reader or read len(buf) (MaxBlockSize) bytes.
-// If one of those is not true, the err will be non-nil.
+// readToEOF will exhaust r or fill buf.
+// If r does not EOF upon reading up to len(buf) bytes then readToEOF will return
+// io.ErrShortBuffer and an additional byte will be discarded from the reader.
 func readToEOF(r io.Reader, buf []byte) (n int, err error) {
 	for err == nil && n < len(buf) {
 		var nn int
 		nn, err = r.Read(buf[n:])
 		n += nn
 	}
-	if err == io.EOF {
-		err = nil
-	} else if err == nil && n != MaxBlockSize {
-		return n, fmt.Errorf("bgzf: expected EOF at end of block after reading %d bytes", n)
+	switch {
+	case err == io.EOF:
+		return n, nil
+	case n == MaxBlockSize && err == nil:
+		// some readers may return quickly when passed a zero-length byte slice.
```

Comment:
The complete text/formating here would be preferable - and even if not, with a leading capital and short over two lines. It is paranoic - the io.Reader docs say that the following call to Read should return a `0, nil`, but I know some that don't.